system/lib/wasmfs/syscalls.cpp

```diff
@@ -426,4 +430,72 @@ __wasi_errno_t __wasi_fd_seek(__wasi_fd_t fd,
 
   return __WASI_ERRNO_SUCCESS;
 }
+
+long __syscall_chdir(long path) {
+  auto pathParts = splitPath((char*)path);
+
+  if (pathParts.empty()) {
+    return -ENOENT;
+  }
+
+  long err;
+  auto dir = getDir(pathParts.begin(), pathParts.end(), err);
+
+  if (!dir) {
+    return err;
+  }
+
+  wasmFS.locked().setCWD(dir);
+  return 0;
+}
+
+long __syscall_getcwd(long buf, long size) {
+  // Check if buf points to a bad address.
+  if (!buf && size > 0) {
+    return -EFAULT;
+  }
+
+  // Check if the size argument is zero and buf is not a null pointer.
+  if (buf && size == 0) {
+    return -EINVAL;
+  }
+
+  auto curr = wasmFS.locked().getCWD();
+
+  std::string result = "";
+
+  while (curr != wasmFS.getRootDirectory()) {
+    auto parent = curr->locked().getParent();
+    // Note: parent.lock() creates a new shared_ptr to the same Directory
+    // specified by the parent weak_ptr.
+    // Check if the parent exists.
```

Comment:
The comment here is not clear to me. Why do we need to create a new `shared_ptr` just to check if the parent exists? I might be wrong but I think we need to lock, as the code correctly does, before checking for existence, but whether that creates a new `shared_ptr` or not seems not relevant, unless I'm missing something.

Reply:
This is an outdated comment, I removed it. But the lock is necessary to get a `shared_ptr` from the `weak_ptr` to parent. Otherwise we cannot access the resource. I moved this comment to the `getParent()` method instead.